### PR TITLE
Mandd/gawithcustomsampler

### LINKED ISSUE #2060: [DEFECT] Optimizers cannot be "initialized" with CustomSampler <Source> Files
### Thank you for the defect report

- [X] I am using the latest version of `RAVEN`.
- [X] I have read the [Wiki](https://github.com/idaholab/raven/wiki).
- [X] I have created a [minimum, reproducible example](https://stackoverflow.com/help/minimal-reproducible-example)
that demonstrates the defect.

### Defect Description

Initial points (for Gradient Descent) and initial population (Genetic algorithm and Simulating annealing) cannot be
set with a ```CustomSampler``` of type ```Files```. E.g. Custom sampler defined as follows:

```
  <Samplers>
    <CustomSampler name="restartSampler">
      <Source class="Files" type="">restart.csv</Source>
      <variable name="x"/>
      <variable name="y"/>
      <variable name="z"/>
    </CustomSampler>
  </Samplers>
```
If used in the Optimizers the following crash is detected:
```
Traceback (most recent call last):
  File "./raven/raven_framework.py", line 25, in <module>
    sys.exit(main(True))
  File "./raven/ravenframework/Driver.py", line 207, in main
    raven()
  File "./raven/ravenframework/Driver.py", line 160, in raven
    simulation.run()
  File "./raven/ravenframework/Simulation.py", line 890, in run
    self.executeStep(stepInputDict, stepInstance)
  File "./raven/ravenframework/Simulation.py", line 823, in executeStep
    stepInstance.takeAstep(stepInputDict)
  File "./raven/ravenframework/Steps/Step.py", line 314, in takeAstep
    self._initializeStep(inDictionary)
  File "./raven/ravenframework/Steps/Step.py", line 247, in _initializeStep
    self._localInitializeStep(inDictionary)
  File "./raven/ravenframework/Steps/MultiRun.py", line 100, in _localInitializeStep
    self._initializeSampler(inDictionary)
  File "./raven/ravenframework/Steps/MultiRun.py", line 72, in _initializeSampler
    inDictionary[self.samplerType].initialize(**self._samplerInitDict)
  File "./raven/ravenframework/Optimizers/GeneticAlgorithm.py", line 387, in initialize
    RavenSampled.initialize(self, externalSeeding=externalSeeding, solutionExport=solutionExport)
  File "./raven/ravenframework/Optimizers/RavenSampled.py", line 160, in initialize
    Optimizer.initialize(self, externalSeeding=externalSeeding, solutionExport=solutionExport)
  File "./raven/ravenframework/Optimizers/Optimizer.py", line 301, in initialize
    self._initializeInitSampler(externalSeeding)
  File "./raven/ravenframework/Optimizers/Optimizer.py", line 406, in _initializeInitSampler
    self._initSampler._localGenerateAssembler(samplerInit)
  File "./raven/ravenframework/Samplers/CustomSampler.py", line 155, in _localGenerateAssembler
    self.assemblerDict[key].append([entity, etype, name, initDict[entity][name]])
KeyError: 'Files'
```

No problems if a DataObject is used in the CustomSampler:

```
  <Samplers>
    <CustomSampler name="restartSampler">
      <Source class="DataObjects" type="PointSet">restart</Source>
      <variable name="x"/>
      <variable name="y"/>
      <variable name="z"/>
    </CustomSampler>
  </Samplers>
```

### Steps to Reproduce

Create a restart case with GA or GD using a custom sampler with ```<Source>``` == ```<Files>```

### Expected Behavior

No crash :D 

### Screenshots and Input Files

_No response_

### OS

Linux

### OS Version

_No response_

### Dependency Manager

PIP

### For Change Control Board: Issue Review

- [x] Is it tagged with a type: defect or task?
- [x] Is it tagged with a priority: critical, normal or minor?
- [x] If it will impact requirements or requirements tests, is it tagged with requirements?
- [x] If it is a defect, can it cause wrong results for users? If so an email needs to be sent to the users.
- [x] Is a rationale provided? (Such as explaining why the improvement is needed or why current code is wrong.)

### For Change Control Board: Issue Closure

- [x] If the issue is a defect, is the defect fixed?
- [x] If the issue is a defect, is the defect tested for in the regression test system?  (If not explain why not.)
- [x] If the issue can impact users, has an email to the users group been written (the email should specify if the defect impacts stable or master)?
- [x] If the issue is a defect, does it impact the latest release branch? If yes, is there any issue tagged with release (create if needed)?
- [x] If the issue is being closed without a pull request, has an explanation of why it is being closed been provided?

### PR DESCRIPTION
**Pull request by mandd**
--------
Pull Request Description
--------
##### What issue does this change request address? (Use "#" before the issue to link it, i.e., #42.)
Closes #2060 

##### What are the significant changes in functionality due to this change request?

----------------
For Change Control Board: Change Request Review
----------------
The following review must be completed by an authorized member of the Change Control Board.
- [x] 1. Review all computer code.
- [x] 2. If any changes occur to the input syntax, there must be an accompanying change to the user manual and xsd schema. If the input syntax change deprecates existing input files, a conversion script needs to be added (see Conversion Scripts).
- [x] 3. Make sure the Python code and commenting standards are respected (camelBack, etc.) - See on the [wiki](https://github.com/idaholab/raven/wiki/RAVEN-Code-Standards#python) for details.
- [x] 4. Automated Tests should pass, including run_tests, pylint, manual building and xsd tests. If there are changes to Simulation.py or JobHandler.py the qsub tests must pass.
- [x] 5. If significant functionality is added, there must  be tests added to check this. Tests should cover all possible options.  Multiple short tests are preferred over one large test. If new development on the internal JobHandler parallel system is performed, a cluster test must be added setting, in <RunInfo> XML block, the node ```<internalParallel>``` to True.
- [x] 6. If the change modifies or adds a requirement or a requirement based test case, the Change Control Board's Chair or designee also needs to approve the change.  The requirements and the requirements test shall be in sync.
- [x] 7. The merge request must reference an issue.  If the issue is closed, the issue close checklist shall be done.
- [x] 8. If an analytic test is changed/added is the the analytic documentation updated/added?
- [x] 9. If any test used as a basis for documentation examples (currently found in `raven/tests/framework/user_guide` and `raven/docs/workshop`) have been changed, the associated documentation must be reviewed and assured the text matches the example.

### CODE REVIEW

**Comment by mandd** [comment]
@aalfonsi this PR addesses your issue #2060

**Comment by aalfonsi** [comment]
Thanks a lot!

File: ravenframework/CodeInterfaceClasses/MooseBasedApp/MooseInputParser.py [hunk 1]
```diff
@@ -106,7 +107,9 @@ def getpotToInputTree(getpot):
       #------------------
       # attributes and values
       elif '=' in line:
+        print('HERE')
```
Comment (wangcj05) [inline ravenframework/CodeInterfaceClasses/MooseBasedApp/MooseInputParser.py:110]: remove?

File: ravenframework/CodeInterfaceClasses/MooseBasedApp/MooseInputParser.py [hunk 2]
```diff
@@ -106,7 +107,9 @@ def getpotToInputTree(getpot):
...
         attribute, value = (x.strip() for x in line.split('=', maxsplit=1))
+        print(attribute, value)
```
Comment (wangcj05) [inline ravenframework/CodeInterfaceClasses/MooseBasedApp/MooseInputParser.py:112]: remove?

File: ravenframework/CodeInterfaceClasses/MooseBasedApp/MooseInputParser.py [hunk 3]
```diff
@@ -124,6 +127,7 @@ def getpotToInputTree(getpot):
             # DO NOT continue, keep going until multiline is closed
             continue
         else:
+          print('HERE1')
```
Comment (wangcj05) [inline ravenframework/CodeInterfaceClasses/MooseBasedApp/MooseInputParser.py:130]: remove?

File: ravenframework/Optimizers/Optimizer.py [hunk 4]
```diff
@@ -208,7 +208,8 @@ def _localGenerateAssembler(self, initDict):
     self.assemblerDict['DataObjects'] = []
     self.assemblerDict['Distributions'] = []
     self.assemblerDict['Functions'] = []
-    for mainClass in ['DataObjects', 'Distributions', 'Functions']:
+    self.assemblerDict['Files'] = []
+    for mainClass in ['DataObjects', 'Distributions', 'Functions', 'Files']:
```
Comment (wangcj05) [inline ravenframework/Optimizers/Optimizer.py:212]: maybe define a validAssembler list which can be used in multiple places.

File: tests/framework/Optimizers/GeneticAlgorithms/discrete/unconstrained/testGAMaxwRep_custom_sampler.xml [hunk 5]
```diff
@@ -0,0 +1,166 @@
...
+      </variable>
+
+      <variable name="x3">
+        <distribution>uniform_dist_Repl_1</distribution>
+      </variable>
+
+      <objective>ans</objective>
+      <TargetEvaluation class="DataObjects" type="PointSet">optOut</TargetEvaluation>
+    <Sampler class="Samplers" type="CustomSampler">restartSampler</Sampler>
+    <!--Sampler class="Samplers" type="MonteCarlo">MC_samp</Sampler-->
```
Comment (wangcj05) [inline tests/framework/Optimizers/GeneticAlgorithms/discrete/unconstrained/testGAMaxwRep_custom_sampler.xml:104]: remove?

File: tests/framework/Optimizers/GeneticAlgorithms/discrete/unconstrained/testGAMaxwRep_custom_sampler.xml [hunk 6]
```diff
@@ -0,0 +1,166 @@
...
+      <variable name="x1">
+        <distribution>uniform_dist_Repl_1</distribution>
+      </variable>
+      <variable name="x2">
+        <distribution>uniform_dist_Repl_1</distribution>
+      </variable>
+      <variable name="x3">
+        <distribution>uniform_dist_Repl_1</distribution>
+      </variable>
+    </MonteCarlo>
```
Comment (wangcj05) [inline tests/framework/Optimizers/GeneticAlgorithms/discrete/unconstrained/testGAMaxwRep_custom_sampler.xml:123]: remove? Not used.

**Review by wangcj05 — CHANGES_REQUESTED** [review]
I have a coupe of minor comments for you to consider. @mandd

File: ravenframework/CodeInterfaceClasses/MooseBasedApp/MooseInputParser.py (hunk 2 above)
Comment (mandd) [reply ravenframework/CodeInterfaceClasses/MooseBasedApp/MooseInputParser.py:112]: Fixed, edit for a different issue

File: ravenframework/CodeInterfaceClasses/MooseBasedApp/MooseInputParser.py (hunk 3 above)
Comment (mandd) [reply ravenframework/CodeInterfaceClasses/MooseBasedApp/MooseInputParser.py:130]: Fixed, edit for a different issue

File: tests/framework/Optimizers/GeneticAlgorithms/discrete/unconstrained/testGAMaxwRep_custom_sampler.xml (hunk 5 above)
Comment (mandd) [reply tests/framework/Optimizers/GeneticAlgorithms/discrete/unconstrained/testGAMaxwRep_custom_sampler.xml:104]: fixed

File: tests/framework/Optimizers/GeneticAlgorithms/discrete/unconstrained/testGAMaxwRep_custom_sampler.xml (hunk 6 above)
Comment (mandd) [reply tests/framework/Optimizers/GeneticAlgorithms/discrete/unconstrained/testGAMaxwRep_custom_sampler.xml:123]: fixed

File: ravenframework/Optimizers/Optimizer.py (hunk 4 above)
Comment (mandd) [reply ravenframework/Optimizers/Optimizer.py:212]: added

File: ravenframework/Optimizers/Optimizer.py [hunk 7]
```diff
@@ -401,7 +405,7 @@ def _initializeInitSampler(self, externalSeeding):
     self._initSampler = sampler
     # initialize sampler
     samplerInit = {}
-    for entity in ['Distributions', 'Functions', 'DataObjects']:
+    for entity in ['Distributions', 'Functions', 'DataObjects','Files']:
```
Comment (wangcj05) [inline ravenframework/Optimizers/Optimizer.py:408]: Could you change the list to use self.optAssemblerList here?

File: ravenframework/Optimizers/Optimizer.py [hunk 8]
```diff
@@ -237,6 +240,7 @@ def _localWhatDoINeed(self):
     needDict['Distributions'] = [(None,'all')]
     needDict['Functions'    ] = [(None,'all')]
     needDict['DataObjects'  ] = [(None,'all')]
+    needDict['Files'        ] = [(None,'all')]
```
Comment (wangcj05) [inline ravenframework/Optimizers/Optimizer.py:243]: It seems we can do a loop over self.optAssemblerList here.

Comment (mandd) [reply ravenframework/Optimizers/Optimizer.py]: good point

File: ravenframework/Optimizers/Optimizer.py [hunk 9]
```diff
@@ -234,9 +237,12 @@ def _localWhatDoINeed(self):
...
     needDict = {}
-    needDict['Distributions'] = [(None,'all')]
-    needDict['Functions'    ] = [(None,'all')]
-    needDict['DataObjects'  ] = [(None,'all')]
+    for elem in self.optAssemblerList:
+      needDict[elem] = [(None,'all')]
+    #needDict['Distributions'] = [(None,'all')]
+    #needDict['Functions'    ] = [(None,'all')]
+    #needDict['DataObjects'  ] = [(None,'all')]
+    #needDict['Files'        ] = [(None,'all')]
```
Comment (wangcj05) [inline ravenframework/Optimizers/Optimizer.py:245]: remove these lines?

Comment (mandd) [reply ravenframework/Optimizers/Optimizer.py:245]: sorry

**Comment by wangcj05** [comment]
Once tests are green, this PR can be merged.

**Comment by wangcj05** [comment]
@mandd I guess there is a consistent failure in Windows, which is caused by test InterpolatedMaxCycles. This test has been recently revived in PR#2035. The purpose of the test is test the validity of input structure for ARMA, it is not made for consistent check.

image

 I recommend to increase the rel_err for this test or we do not check the consistency in the csv file, just checkout it produces csv files.

**Comment by joshua-cogliati-inl** [comment]
> @mandd I guess there is a consistent failure in Windows, which is caused by test InterpolatedMaxCycles. This test has been recently revived in PR#2035. The purpose of the test is test the validity of input structure for ARMA, it is not made for consistent check.
> 
> image
> 
> I recommend to increase the rel_err for this test or we do not check the consistency in the csv file, just checkout it produces csv files.

I agree, change "csv =" to "output =" in this test.

**Review by wangcj05 — APPROVED** [review]

**Comment by wangcj05** [comment]
PR looks good and checklist is satisfied.

**Pull request merged**